guide_drafts/model-derives.md

```diff
@@ -0,0 +1,926 @@
+
+```rust
+error[E0599]: no method named `id` found for type `diesel_demo_cli::models::User` in the current scope
+  --> src/bin/main.rs:37:48
+   |
+37 |     println!("Our Hero's Id: {}", hero.id());
+   |                                        ^^ field, not a method
+   |
+   = help: did you mean to write `hero.id` instead of `hero.id(...)`?
+  --> src/bin/main.rs:39:5
+   |
+39 |     diesel::update(&hero).set((
+   |     ^^^^^^^^^^^^^^ the trait `diesel::associations::HasTable` is not implemented for `diesel_demo_cli::models::User`
+   |
+   = note: required because of the requirements on the impl of `diesel::associations::HasTable` for `&diesel_demo_cli::models::User`
+   = note: required because of the requirements on the impl of `diesel::query_builder::IntoUpdateTarget` for `&diesel_demo_cli::models::User`
+   = note: required by `diesel::update`
+
+error[E0277]: the trait bound `&diesel_demo_cli::models::User: diesel::Identifiable` is not satisfied
+  --> src/bin/main.rs:39:5
+   |
+39 |     diesel::update(&hero).set((
+   |     ^^^^^^^^^^^^^^ the trait `diesel::Identifiable` is not implemented for `&diesel_demo_cli::models::User`
+   |
+   = note: required because of the requirements on the impl of `diesel::query_builder::IntoUpdateTarget` for `&diesel_demo_cli::models::User`
+   = note: required by `diesel::update`
+```
+
+From this error,
+you can see the compiler assumes we are looking for a field on the struct,
+but what we want is just the value of the primary key,
+which happens to also be the `id` field.
+The compiler is also giving us some hints about trait bounds in each one of those *note* sections.
+The first two notes about `HasTable` and `IntoUpdateTarget` might be helpful if we refer to the [Diesel API docs](http://docs.diesel.rs/diesel/associations/trait.Identifiable.html),
+but the third error lets us know exactly which trait isn't satisfied! :smile:
+
+The following code is generated by `cargo expand` to get our example code working again.
+
+```rust
+#[derive(Queryable)]
+pub struct User {
+    pub id: i32,
+    pub first_name: String,
+    pub last_name: String,
+    pub email: String,
+}
+
+impl <'ident> Identifiable for &'ident User {
+    type Id = (&'ident i32);
+
+    fn id(self) -> Self::Id {
+        (&self.id)
+    }
+}
+
+impl HasTable for User {
+    type Table = users::table;
+
+    fn table() -> Self::Table {
+        users::table
+    }
+}
+```
+
+## <a name="aschangeset"></a>AsChangeset
+
+For more slightly more complicated database updates,
+`AsChangeset` can make your development easier.
+In this section we will change our schema to make our `email` field nullable.
+The nullable fields on our model structs will now be of type `Option<T>`.
+You will probably be not changing the primary key of the row or rows that you're updating.
+For this reason, we probably won't have `AsChangeset` and `Queryable` annotated on the same struct,
+which means we'll need another struct for updating our records.
```

Comment:
Maybe even add this somewhere? I don't think we ever wrote that down somewhere prominently:

> It's considered good practice in diesel to have one struct per "usage". E.g., one struct to query users, one to create new users, one to change just the email, etc. This often corresponds to the way applications access the database. Don't be afraid to have multiple structs per database table—they are not like the classes that correspond to tables 1:1 in other ORMs.

Reply:
I think this is a really important point. I'm coming from a rails background, so this seemed alien at first. I'll think about a good spot to add it.